src/types.rs

```diff
@@ -475,3 +506,209 @@ impl<'buf> RecvMsgOut<'buf> {
+enum MatchMultiple {
+    /// Match all requests. This indiscriminately cancels all requests,
+    /// regardless of user_data or fd.
+    All,
+    /// Match any request which matches the user_data or fd.
+    Any,
+    /// Match only a single request. This is the default behavior.
+    Single,
+}
+
+impl MatchOn {
+    /// Match a single request with the given user_data field.
+    ///
+    /// ### Example
+    ///
+    /// ```
+    /// use io_uring::types;
+    ///
+    /// // Match a single request with user_data 42
+    /// types::MatchOn::user_data(42);
+    /// ```
+    pub fn user_data(user_data: u64) -> Self {
+        Self {
+            user_data: Some(user_data),
+            fd: None,
+            multiple: MatchMultiple::Single,
+        }
+    }
+
+    /// Match a single request using the given file descriptor.
+    ///
+    /// ### Availability
+    ///
+    /// - Matching on [`Fd`](types::Fd) is available since 5.19.
+    /// - Matching on [`Fixed`](types::Fixed) is available since 6.0.
+    ///
+    /// ### Example
+    ///
+    /// ```
+    /// use io_uring::types;
+    /// // Match a single request using file descriptor 4
+    /// types::MatchOn::fd(types::Fd(4));
+    ///
+    /// // Match a single request using the fixed file descriptor 8
+    /// types::MatchOn::fd(types::Fixed(8));
+    /// ```
+    pub fn fd(fd: impl sealed::UseFixed) -> Self {
+        Self {
+            user_data: None,
+            fd: Some(fd.into()),
+            multiple: MatchMultiple::Single,
+        }
+    }
+
+    /// Match all requests, regardless of user_data or fd.
+    ///
+    /// If matching all requests, other match criteria are ignored.
+    ///
+    /// ### Example
+    ///
+    /// ```
+    /// use io_uring::types;
+    /// /// Match all in-fliht requests
+    /// types::MatchOn::all();
+    /// ```
+    pub fn all() -> Self {
+        Self {
+            user_data: None,
+            fd: None,
+            multiple: MatchMultiple::All,
+        }
+    }
+
+    /// Apply the match criteria to *any* request.
+    ///
+    /// This will match multiple in-flight requests.
+    ///
+    /// ### Example
+    ///
+    /// ```
+    /// use io_uring::types;
+    /// // Match any request where the user_data or fd is 42.
+    /// types::MatchOn::user_data(42).and_fd(types::Fixed(42)).match_any();
+    /// ```
+    pub fn match_any(mut self) -> Self {
+        self.multiple = MatchMultiple::Any;
+        self
+    }
+
+    /// Additionally, match a request with the given user_data field.
+    ///
+    /// ### Example
+    ///
+    /// ```
+    /// use io_uring::types;
+    /// // Match a single request where the user_data is
+    /// // 42 or the fd is 4.
+    /// types::MatchOn::user_data(42).and_fd(types::Fixed(4));
+    /// ```
+    ///
+    /// Note: This will overwrite any previous user_data match criteria.
+    pub fn and_fd(mut self, fd: impl sealed::UseFixed) -> Self {
+        if matches!(self.multiple, MatchMultiple::All) {
```

Comment:
I don't follow the reasoning behind this test for comparison to All. One would still want the ability to match against all for a given fd.